File: experiments/runs/2026-09-16_08-12-47_claim-office-example-mapping_exact-sol-v1.6-test-list-dimensions-cc_opus-5-no-thinking/src/reimbursement.ts
import type { Item } from "./item.js";
import { multiply, rational, type Rational } from "./rational.js";

/** Reimbursement clauses decide how much of a damage amount MHPCO covers. */
const REDUCED_REIMBURSEMENT_PERCENT = 50;
const PERCENT = 100;
const REDUCED_REIMBURSEMENT_LEVEL = 8;

function isDeeplyEnchanted(item: Item): boolean {
  return (item.enchantment ?? 0) >= REDUCED_REIMBURSEMENT_LEVEL;
}

export function reimbursableAmount(item: Item, damageAmountInG: number): Rational {
  const damage = rational(damageAmountInG);
  if (isDeeplyEnchanted(item)) {
    return multiply(damage, rational(REDUCED_REIMBURSEMENT_PERCENT, PERCENT));
  }
  return damage;
}

File: experiments/runs/2026-09-20_10-40-56_claim-office-java-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking/src/main/java/Reimbursement.java
import java.math.BigDecimal;
import java.util.List;

/**
 * What the MHPCO reimburses for an incident, before the policy's cap is
 * applied.
 *
 * <p>Knows that a deductible applies per damage event — once for each damaged
 * item — and that it is taken off after the clauses have said how much of the
 * damage is reimbursed at all. Defers which item a damage entry names, and
 * what the clauses make of it, to those two.
 */
final class Reimbursement {

    private static final BigDecimal DEDUCTIBLE_IN_G = BigDecimal.valueOf(100);

    private Reimbursement() {
    }

    static BigDecimal of(List<Item> insured, Incident incident) {
        BigDecimal reimbursement = BigDecimal.ZERO;
        for (DamagedItem damaged : DamagedItem.of(insured, incident.damages())) {
            reimbursement = reimbursement.add(reimbursementFor(damaged));
        }
        return reimbursement;
    }

    private static BigDecimal reimbursementFor(DamagedItem damaged) {
        BigDecimal damage = BigDecimal.valueOf(damaged.damage().amount());
        return DamageClause.reimbursedPartOf(damaged.insured(), damage)
                .subtract(DEDUCTIBLE_IN_G)
                .max(BigDecimal.ZERO);
    }
}
